code/crates/engine/src/consensus.rs

```diff
@@ -1115,76 +1133,86 @@ where
+                        }
                     }
-                }
 
-                self.network
-                    .cast(NetworkMsg::PublishLivenessMsg(msg))
-                    .map_err(|e| eyre!("Error when broadcasting liveness message: {e:?}"))?;
+                    self.network
+                        .cast(NetworkMsg::PublishLivenessMsg(msg))
+                        .map_err(|e| eyre!("Error when broadcasting liveness message: {e:?}"))?;
+                }
 
                 Ok(r.resume_with(()))
             }
 
             Effect::RepublishVote(msg, r) => {
-                // Notify any subscribers that we are about to rebroadcast a vote
-                self.tx_event.send(|| Event::RepublishVote(msg.clone()));
+                if is_active_validator {
+                    // Notify any subscribers that we are about to rebroadcast a vote
+                    self.tx_event.send(|| Event::RepublishVote(msg.clone()));
 
-                self.network
-                    .cast(NetworkMsg::PublishLivenessMsg(LivenessMsg::Vote(msg)))
-                    .map_err(|e| eyre!("Error when rebroadcasting vote message: {e:?}"))?;
+                    self.network
+                        .cast(NetworkMsg::PublishLivenessMsg(LivenessMsg::Vote(msg)))
+                        .map_err(|e| eyre!("Error when rebroadcasting vote message: {e:?}"))?;
+                }
 
                 Ok(r.resume_with(()))
             }
 
             Effect::RepublishRoundCertificate(certificate, r) => {
-                // Notify any subscribers that we are about to rebroadcast a round certificate
-                self.tx_event
-                    .send(|| Event::RebroadcastRoundCertificate(certificate.clone()));
-
-                self.network
-                    .cast(NetworkMsg::PublishLivenessMsg(
-                        LivenessMsg::SkipRoundCertificate(certificate),
-                    ))
-                    .map_err(|e| {
-                        eyre!("Error when rebroadcasting round certificate message: {e:?}")
-                    })?;
+                if is_active_validator {
+                    // Notify any subscribers that we are about to rebroadcast a round certificate
+                    self.tx_event
+                        .send(|| Event::RebroadcastRoundCertificate(certificate.clone()));
+
+                    self.network
+                        .cast(NetworkMsg::PublishLivenessMsg(
+                            LivenessMsg::SkipRoundCertificate(certificate),
+                        ))
+                        .map_err(|e| {
+                            eyre!("Error when rebroadcasting round certificate message: {e:?}")
+                        })?;
+                }
 
                 Ok(r.resume_with(()))
             }
 
             Effect::GetValue(height, round, timeout, r) => {
-                let timeout_duration = state.timeouts.duration_for(timeout.kind);
+                if is_active_validator {
```

Comment:
In the original algorithm, Tendermint _knows_ when it's the proposer. There's an `if` distinguishing both cases. Is that "if" now extracted to the engine?

If not, then, shouldn't this "if" rather be a sanity check?

Reply:
No, core-consensus still does the proposer check and will not emit a `GetValue` effect if it's not the proposer. So this check is more of a sanity check and could be removed altogether.

Reply:
I see. Then, my comment is that the "if" you've added should become an `assert` or whatever provision you have in place to detect internal bugs.

Reply:
this is for the case where a validator is supposed to participate in consensus but it is far behind the tip of the chain.  So consensus (who doesn't know that) says `GetValue` but the engine filters it.

Reply:
Ok, it's not clear (at least to me) that this condition is for sync-only mode. So maybe adding a comment with that extra info for whoever reads that code would help.
You can disregard my comment about the "if" becoming a sanity check, as I wasn't aware of the assumption this is his on sync-only mode. I was having normal consensus operation in mind (so only up-to-date nodes participating).